src/gui/torrentoptionsdialog.cpp

```diff
@@ -0,0 +1,331 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "torrentoptionsdialog.h"
+
+#include <QMessageBox>
+
+#include "base/bittorrent/session.h"
+#include "base/bittorrent/torrenthandle.h"
+#include "ui_torrentoptionsdialog.h"
+#include "utils.h"
+
+TorrentOptionsDialog::TorrentOptionsDialog(QWidget *parent, const QVector<BitTorrent::TorrentHandle *> selectedTorrents)
+    : QDialog(parent)
+    , m_selectedTorrents(selectedTorrents)
+    , m_ui(new Ui::TorrentOptionsDialog)
+{
+    m_ui->setupUi(this);
+
+    m_session = BitTorrent::Session::instance();
+
+    connect(m_ui->sliderUploadLimit, &QSlider::valueChanged, this, &TorrentOptionsDialog::updateUploadSpinValue);
+    connect(m_ui->sliderDownloadLimit, &QSlider::valueChanged, this, &TorrentOptionsDialog::updateDownloadSpinValue);
+    connect(m_ui->spinUploadLimit, qOverload<int>(&QSpinBox::valueChanged)
+            , this, &TorrentOptionsDialog::updateUploadSliderValue);
+    connect(m_ui->spinDownloadLimit, qOverload<int>(&QSpinBox::valueChanged)
+            , this, &TorrentOptionsDialog::updateDownloadSliderValue);
+
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 15, 0))
+    connect(m_ui->buttonGroup, &QButtonGroup::idClicked
+            , this, &TorrentOptionsDialog::handleRatioTypeChanged);
+#else
+    connect(m_ui->buttonGroup, qOverload<int>(&QButtonGroup::buttonClicked)
+            , this, &TorrentOptionsDialog::handleRatioTypeChanged);
+#endif
+    connect(m_ui->checkMaxRatio, &QCheckBox::toggled, this, &TorrentOptionsDialog::enableRatioSpin);
+    connect(m_ui->checkMaxTime, &QCheckBox::toggled, this, &TorrentOptionsDialog::enableTimeSpin);
+
+    handleRatioTypeChanged();
+
+    initDialog();
+
+    Utils::Gui::resize(this);
+}
+
+TorrentOptionsDialog::~TorrentOptionsDialog()
+{
+    delete m_ui;
+}
+
+void TorrentOptionsDialog::accept()
+{
+    if (m_ui->radioTorrentLimit->isChecked() && !m_ui->checkMaxRatio->isChecked() && !m_ui->checkMaxTime->isChecked())
+    {
+        QMessageBox::critical(this, tr("No share limit method selected"), tr("Please select a limit method first"));
```

Comment:
Critical message is too much, IMO.

Reply:
Also I would just disable OK button if control values have inconsistent states.

Reply:
>Critical message is too much, IMO.

Note that i just moved this here. About critical: which of the other options convey "You can't proceed until..."?

>Also I would just disable OK button if control values have inconsistent states.

If it was right after performing an invalid action i'd agree but here it's because an extra action is needed.

Reply:
Well, I guess you're right. Don't stop there.